session.go

```diff
@@ -0,0 +1,115 @@
+package mockoidc
+
+import (
+	"strings"
+	"time"
+
+	"github.com/dgrijalva/jwt-go"
+)
+
+// Session is a struct that maintains session state across requests
+type Session struct {
+	SessionID  string
+	Scopes     []string
+	OAuthState string
+	OIDCNonce  string
+	User       *User
+}
+
+// SessionStore is a map of #Session structs
+type SessionStore struct {
+	Store map[string]*Session
+}
+
+// NewSessionStore initializes the SessionStore for this server
+func NewSessionStore() *SessionStore {
+	var ss SessionStore
+	ss.Store = make(map[string]*Session)
+	return &ss
+}
+
+// NewSession returns a pointer to a new Session struct
+func (ss *SessionStore) NewSession(scope string, oAuthState string, oidcNonce string, user *User) (*Session, error) {
+	sessionID, err := nonce(24)
+	if err != nil {
+		return nil, err
+	}
+
+	session := &Session{
+		SessionID:  sessionID,
+		Scopes:     strings.Split(scope, " "),
+		OAuthState: oAuthState,
+		OIDCNonce:  oidcNonce,
+		User:       user,
+	}
+
+	ss.Store[sessionID] = session
+
+	return session, nil
+}
+
+// AccessToken returns the JWT token with the appropriate claims for an access token
+func (s *Session) AccessToken(config *Config, kp *Keypair, now time.Time) (string, error) {
+
+	claims := s.standardClaims(config, config.AccessTTL, now)
+
+	return kp.SignJWT(claims)
+}
+
+// standardClaims returns a populated jwt.StandardCLaims struct
+func (s *Session) standardClaims(config *Config, ttl int, now time.Time) jwt.StandardClaims {
+
+	return jwt.StandardClaims{
+		Audience:  config.ClientID,
+		ExpiresAt: now.Unix() + int64(ttl),
+		Id:        s.SessionID,
+		IssuedAt:  now.Unix(),
+		Issuer:    config.Issuer,
+		NotBefore: now.Unix(),
+		Subject:   s.User.Email,
+	}
+}
+
+// RefreshToken returns the JWT token with the appropriate claims for an access token
+func (s *Session) RefreshToken(config *Config, kp *Keypair, now time.Time) (string, error) {
+	claims := s.standardClaims(config, config.RefreshTTL, now)
+
+	return kp.SignJWT(claims)
+}
+
+// IDTokenClaims are the claims to be provided to jwt.Sign to create the ID token
+type IDTokenClaims struct {
+	PreferredUsername string   `json:"preferred_username,omitempty"`
+	Email             string   `json:"email,omitempty"`
+	Phone             string   `json:"phone_number,omitempty"`
+	Address           string   `json:"address,omitempty"`
+	Groups            []string `json:"groups,omitempty"`
+	EmailVerified     bool     `json:"email_verified,omitempty"`
+	Nonce             string   `json:nonce,omitempty`
+	jwt.StandardClaims
+}
+
+// IDToken returns the JWT token with the appropriate claims for an access token
+func (s *Session) IDToken(config *Config, kp *Keypair, now time.Time) (string, error) {
+
+	idClaims := IDTokenClaims{
+		StandardClaims: s.standardClaims(config, config.AccessTTL, now),
+	}
+
+	idClaims = s.User.IDClaimsForScopes(s.Scopes, idClaims)
+
+	if contains(s.Scopes, "nonce") {
```

Comment:
`nonce` won't be a scope -- it will be in the session if we receive it as a parameter on the initial `/authorize` call (`nonce` and `scopes` are distinct querystring parameters in that request).

So I think here we just set the `nonce` claim if the session's nonce is not `""`